File: openfeature-provider/java/src/test/java/com/spotify/confidence/sdk/ProviderTelemetryResolverTest.java
package com.spotify.confidence.sdk;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import com.spotify.confidence.sdk.flags.resolver.v1.ApplyFlagsRequest;
import com.spotify.confidence.sdk.flags.resolver.v1.RegisterResolveRequest;
import com.spotify.confidence.sdk.flags.resolver.v1.ResolveProcessRequest;
import com.spotify.confidence.sdk.flags.resolver.v1.ResolveProcessResponse;
import com.spotify.confidence.sdk.flags.resolver.v1.Sdk;
import com.spotify.confidence.sdk.flags.resolver.v1.SdkId;
import com.spotify.confidence.sdk.flags.resolver.v1.WriteFlagLogsRequest;
import java.util.ArrayList;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;
import org.junit.jupiter.api.Test;

class ProviderTelemetryResolverTest {
  private static final Sdk SDK =
      Sdk.newBuilder().setId(SdkId.SDK_ID_JAVA_LOCAL_PROVIDER).setVersion("test-version").build();

  @Test
  void emitsOnceAcrossPooledFlushes() {
    final var captured = new ArrayList<WriteFlagLogsRequest>();
    final var resolver =
        new ProviderTelemetryResolver(
            captured::add,
            SDK,
            Map.of("encryption", "true"),
            sink -> new TelemetryTestResolver(sink, 3));

    resolver.flushAllLogs();

    assertThat(providerInitEventCount(captured)).isEqualTo(1);
    final var telemetry = captured.get(0).getTelemetryData();
    assertThat(telemetry.getSdk()).isEqualTo(SDK);
    assertThat(telemetry.getProviderInitRate(0).getLabelsMap()).containsEntry("encryption", "true");
  }

  @Test
  void closeEmitsWithoutResolve() {
    final var captured = new ArrayList<WriteFlagLogsRequest>();
    final var resolver =
        new ProviderTelemetryResolver(
            captured::add, SDK, Map.of(), sink -> new TelemetryTestResolver(sink, 0));

    resolver.close();

    assertThat(providerInitEventCount(captured)).isEqualTo(1);
  }

  @Test
  void retriesAfterSinkFailure() {
    final var attempts = new AtomicInteger();
    final var captured = new ArrayList<WriteFlagLogsRequest>();
    final var resolver =
        new ProviderTelemetryResolver(
            request -> {
              if (attempts.getAndIncrement() == 0) {
                throw new IllegalStateException("send failed");
              }
              captured.add(request);
            },
            SDK,
            Map.of(),
            sink -> new TelemetryTestResolver(sink, 1));

    assertThatThrownBy(resolver::flushAllLogs).isInstanceOf(IllegalStateException.class);
    resolver.flushAllLogs();

    assertThat(providerInitEventCount(captured)).isEqualTo(1);
  }

  private static int providerInitEventCount(ArrayList<WriteFlagLogsRequest> requests) {
    return requests.stream()
        .mapToInt(request -> request.getTelemetryData().getProviderInitRateCount())
        .sum();
  }

  private static final class TelemetryTestResolver implements LocalResolver {
    private final Consumer<WriteFlagLogsRequest> sink;
    private final int flushCount;

    private TelemetryTestResolver(Consumer<WriteFlagLogsRequest> sink, int flushCount) {
      this.sink = sink;
      this.flushCount = flushCount;
    }

    @Override
    public void setResolverState(byte[] state, String accountId, Sdk sdk) {}

    @Override
    public CompletionStage<ResolveProcessResponse> resolveProcess(ResolveProcessRequest request) {
      return CompletableFuture.completedFuture(ResolveProcessResponse.getDefaultInstance());
    }

    @Override
    public void applyFlags(ApplyFlagsRequest request) {}

    @Override
    public void registerResolve(RegisterResolveRequest request) {}

    @Override
    public void flushAllLogs() {
      for (int i = 0; i < flushCount; i++) {
        sink.accept(WriteFlagLogsRequest.getDefaultInstance());
      }
    }

    @Override
    public void flushAssignLogs() {}

    @Override
    public void close() {}

    @Override
    public String prometheusSnapshot() {
      return "";
    }
  }
}

File: openfeature-provider/go/confidence/provider_telemetry_resolver_test.go
package confidence

import (
	"context"
	"testing"

	lr "github.com/spotify/confidence-resolver/openfeature-provider/go/confidence/internal/local_resolver"
	resolvertypes "github.com/spotify/confidence-resolver/openfeature-provider/go/confidence/internal/proto/resolver"
	resolverv1 "github.com/spotify/confidence-resolver/openfeature-provider/go/confidence/internal/proto/resolverinternal"
	"github.com/spotify/confidence-resolver/openfeature-provider/go/confidence/internal/proto/wasm"
)

func TestProviderTelemetryResolverEmitsOnceAcrossPooledFlushes(t *testing.T) {
	var captured []*resolverv1.WriteFlagLogsRequest
	resolver := newProviderTelemetryResolver(
		func(logs *resolverv1.WriteFlagLogsRequest) { captured = append(captured, logs) },
		map[string]string{"encryption": "true"},
		func(sink lr.LogSink) lr.LocalResolver { return &telemetryTestResolver{sink: sink, flushCount: 3} },
	)

	if err := resolver.FlushAllLogs(); err != nil {
		t.Fatal(err)
	}

	if got := providerInitEventCount(captured); got != 1 {
		t.Fatalf("expected one provider init event across pooled flushes, got %d", got)
	}
	telemetry := captured[0].GetTelemetryData()
	if telemetry.GetSdk().GetId() != resolvertypes.SdkId_SDK_ID_GO_LOCAL_PROVIDER {
		t.Fatalf("unexpected SDK ID: %v", telemetry.GetSdk().GetId())
	}
	if telemetry.GetSdk().GetVersion() != Version {
		t.Fatalf("unexpected SDK version: %q", telemetry.GetSdk().GetVersion())
	}
}

func TestProviderTelemetryResolverCloseEmitsWithoutResolve(t *testing.T) {
	var captured []*resolverv1.WriteFlagLogsRequest
	resolver := newProviderTelemetryResolver(
		func(logs *resolverv1.WriteFlagLogsRequest) { captured = append(captured, logs) },
		map[string]string{"encryption": "true"},
		func(sink lr.LogSink) lr.LocalResolver { return &telemetryTestResolver{sink: sink} },
	)

	if err := resolver.Close(context.Background()); err != nil {
		t.Fatal(err)
	}

	if got := providerInitEventCount(captured); got != 1 {
		t.Fatalf("expected shutdown to emit one provider init event, got %d", got)
	}
}

func TestProviderTelemetryResolverOwnsSingleInitSample(t *testing.T) {
	resolver := &providerTelemetryResolver{
		labels: map[string]string{"encryption": "true"},
		sdk:    &resolvertypes.Sdk{Version: "test-version"},
	}
	logs := &resolverv1.WriteFlagLogsRequest{
		TelemetryData: &resolverv1.TelemetryData{
			ProviderInitRate: []*resolverv1.TelemetryData_ProviderInitRate{
				{Count: 1, Labels: map[string]string{"existing": "true"}},
			},
		},
	}

	resolver.addInitTelemetry(logs)

	got := logs.GetTelemetryData().GetProviderInitRate()
	if len(got) != 1 {
		t.Fatalf("expected exactly one provider init sample, got %d", len(got))
	}
	if got[0].GetLabels()["encryption"] != "true" {
		t.Fatalf("expected provider-owned labels, got %v", got[0].GetLabels())
	}
}

func TestProviderTelemetryResolverRetriesAfterSinkFailure(t *testing.T) {
	attempts := 0
	var captured []*resolverv1.WriteFlagLogsRequest
	resolver := newProviderTelemetryResolver(
		func(logs *resolverv1.WriteFlagLogsRequest) {
			attempts++
			if attempts == 1 {
				panic("send failed")
			}
			captured = append(captured, logs)
		},
		nil,
		func(sink lr.LogSink) lr.LocalResolver { return &telemetryTestResolver{sink: sink, flushCount: 1} },
	)

	func() {
		defer func() { _ = recover() }()
		_ = resolver.FlushAllLogs()
	}()
	if err := resolver.FlushAllLogs(); err != nil {
		t.Fatal(err)
	}

	if got := providerInitEventCount(captured); got != 1 {
		t.Fatalf("expected init telemetry to be retried, got %d events", got)
	}
}

func providerInitEventCount(requests []*resolverv1.WriteFlagLogsRequest) int {
	count := 0
	for _, request := range requests {
		count += len(request.GetTelemetryData().GetProviderInitRate())
	}
	return count
}

type telemetryTestResolver struct {
	sink       lr.LogSink
	flushCount int
}

func (r *telemetryTestResolver) SetResolverState(*wasm.SetResolverStateRequest) error { return nil }
func (r *telemetryTestResolver) ResolveProcess(*wasm.ResolveProcessRequest) (*wasm.ResolveProcessResponse, error) {
	return &wasm.ResolveProcessResponse{}, nil
}
func (r *telemetryTestResolver) RegisterResolve(*wasm.RegisterResolveRequest)      {}
func (r *telemetryTestResolver) ApplyFlags(*resolvertypes.ApplyFlagsRequest) error { return nil }
func (r *telemetryTestResolver) FlushAllLogs() error {
	for range r.flushCount {
		r.sink(&resolverv1.WriteFlagLogsRequest{})
	}
	return nil
}
func (r *telemetryTestResolver) FlushAssignLogs() error                 { return nil }
func (r *telemetryTestResolver) PrometheusSnapshot(uint32, bool) string { return "" }
func (r *telemetryTestResolver) Close(context.Context) error            { return nil }
